File: takengine/takengine/src/main/cpp/raster/tilepyramid/LegacyTilePyramidTileReader.cpp

#include "string/StringHacks.hh"

#include "core/GeoPoint.h"
#include "feature/Geometry.h"
#include "raster/tilepyramid/LegacyTilePyramidTileReader.h"
#include "raster/tilepyramid/Tilesetinfo.h"
#include "raster/gdal/GdalLibrary.h"
#include "renderer/BitmapFactory.h"
#include "renderer/BitmapCompositor.h"
#include "math/Utils.h"
#include "renderer/GLRendererGlobals.h"

#define DEBUG_TILE_DRAW 0

#define RESTRICT_COMPOSITE 1

using namespace atakmap;
using namespace atakmap::core;
using namespace atakmap::feature;
using namespace atakmap::raster;
using namespace atakmap::raster::tilepyramid;

using namespace TAK::Engine::Thread;
using namespace TAK::Engine::Util;

LegacyTilePyramidTileReader::RegisterValue::~RegisterValue() {
    if (this->bitmapLoaderRelease) {
        this->bitmapLoaderRelease(this->bitmapLoader);
    }
    if (this->tilesetInfoRelease) {
        this->tilesetInfoRelease(this->tilesetInfo);
    }
}

Mutex LegacyTilePyramidTileReader::staticMutex;
LegacyTilePyramidTileReader::RegisterMap LegacyTilePyramidTileReader::registerMap;

atakmap::raster::tilereader::TileReader *LegacyTilePyramidTileReader::Spi::create(const char *uri, const tilereader::TileReaderFactory::Options *options) {
    
    if (strstr(uri, "tsinfo://") != uri) {
        return NULL;
    }
    
    RegisterMap::iterator entry = registerMap.find(uri + 9);
    if (entry == registerMap.end()) {
        return NULL;
    }
    
    TilesetInfo *tsInfo = entry->second.tilesetInfo;
    if (tsInfo == NULL) {
        return NULL;
    }
    
    TilesetSupport *support = TilesetSupport::create(tsInfo, entry->second.bitmapLoader);
    if(support == NULL)
        return NULL;
    
    PGSC::RefCountableIndirectPtr<TileReader::AsynchronousIO> asyncIO = atakmap::raster::gdal::GdalLibrary::getMasterIOThread();
    if (options && options->asyncIO.get())
            asyncIO = options->asyncIO;
        
    return new LegacyTilePyramidTileReader(tsInfo, support, asyncIO);
}


const double LegacyTilePyramidTileReader::EPSILON = 0.000000001;

LegacyTilePyramidTileReader::LegacyTilePyramidTileReader(atakmap::raster::tilepyramid::TilesetInfo *info, renderer::AsyncBitmapLoader *loader)
: LegacyTilePyramidTileReader(info, TilesetSupport::create(info, loader), atakmap::raster::gdal::GdalLibrary::getMasterIOThread())
{ }

LegacyTilePyramidTileReader::LegacyTilePyramidTileReader(TilesetInfo *info,
                                                         TilesetSupport *support, PGSC::RefCountableIndirectPtr<AsynchronousIO> io)
: tilereader::TileReader(info->getInfo()->getURI(), NULL, INT_MAX, io), support(support) {
    
    if(this->support == NULL) {
        throw std::invalid_argument("TilesetSupport cannot be NULL");
    }
    
    this->support->init();
    this->support->start();
    
    this->tileWidth = info->getTilePixelWidth();
    this->tileHeight = info->getTilePixelHeight();
    
    this->levelCount = info->getLevelCount();
    
    const double zeroHeight = info->getZeroHeight();
    const double zeroWidth = info->getZeroWidth();
    
    const double gridOriginLat = info->getGridOriginLat();
    const double gridOriginLng = info->getGridOriginLng();
    
    feature::Envelope mbb = info->getInfo()->getCoverage(NULL)->getEnvelope();
    
    double south = mbb.minY;
    double north = mbb.maxY;
    double west = mbb.minX;
    double east = mbb.maxX;
    
    // Shrink down on the 4 corners as much as we can
    GeoPoint sw(mbb.minY, mbb.minX);
    GeoPoint nw(mbb.maxY, mbb.minX);
    GeoPoint ne(mbb.maxY, mbb.maxX);
    GeoPoint se(mbb.minY, mbb.maxX);
    
    const double coverageSouth = fmin(sw.latitude, se.latitude);
    const double coverageNorth = fmax(nw.latitude, ne.latitude);
    const double coverageEast = fmax(ne.longitude, se.longitude);
    const double coverageWest = fmin(nw.longitude, sw.longitude);
    
    south = _alignMin(gridOriginLat, fmax(coverageSouth, south), zeroHeight);
    west = _alignMin(gridOriginLng, fmax(coverageWest, west), zeroWidth);
    north = _alignMax(gridOriginLat, fmin(coverageNorth, north), zeroHeight);
    east = _alignMax(gridOriginLng, fmin(coverageEast, east), zeroWidth);
    
    int _gridHeight = info->getGridHeight();
    if (_gridHeight < 0)
        _gridHeight = (int) (((north - south) + EPSILON) / zeroHeight);
    int _gridWidth = info->getGridWidth();
    if (_gridWidth < 0)
        _gridWidth = (int) (((east - west) + EPSILON) / zeroWidth);
    int _gridX = info->getGridOffsetX();
    if (_gridX < 0)
        _gridX = (int) ((west - gridOriginLng + EPSILON) / zeroWidth);
    int _gridY = info->getGridOffsetY();
    if (_gridY < 0)
        _gridY = (int) ((south - gridOriginLat + EPSILON) / zeroHeight);
    
    this->gridHeight = _gridHeight;
    this->gridWidth = _gridWidth;
    this->gridOffsetX = _gridX;
    this->gridOffsetY = _gridY;
    
    this->width = this->tileWidth*(this->gridWidth<<(this->levelCount-1));
    this->height = this->tileHeight*(this->gridHeight<<(this->levelCount-1));
}

LegacyTilePyramidTileReader::~LegacyTilePyramidTileReader() { }

void LegacyTilePyramidTileReader::start() {
    this->support->start();
}

void LegacyTilePyramidTileReader::stop() {
    this->support->stop();
}

/**************************************************************************/
// Tile Reader

void LegacyTilePyramidTileReader::disposeImpl() {
    TileReader::disposeImpl();
    if(this->support != NULL) {
        this->support->release();
        this->support = NULL;
    }
}

static const uint64_t NANOS_PER_USEC = 1000ULL;
static const uint64_t NANOS_PER_MILLISEC = 1000ULL * NANOS_PER_USEC;
static const uint64_t NANOS_PER_SEC = 1000ULL * NANOS_PER_MILLISEC;

void LegacyTilePyramidTileReader::cancelAsyncRead(int id) {
    TileReader::cancelAsyncRead(id);
    
    if(this->pending.valid())
        this->pending.cancel();
}

LegacyTilePyramidTileReader::ReadResult
LegacyTilePyramidTileReader::getTileImpl(int level, int64_t tileColumn, int64_t tileRow, renderer::Bitmap *retval) {
    
    if (level < 0)
        throw std::invalid_argument("level");
    
    // need to invert level and tile row for lower-left origin of legacy
    // tileset infrastructure
    const int tsLevel = (this->levelCount-level-1);
    tileRow = (this->gridHeight<<tsLevel)-tileRow-1;
    
    const int latIndex = (int)tileRow + (this->gridOffsetY<<tsLevel);
    const int lngIndex = (int)tileColumn + (this->gridOffsetX<<tsLevel);
    
    /*TODO--const BitmapFactory.Options opts = new BitmapFactory.Options();
    opts.inPreferredConfig = Bitmap.Config.ARGB_8888;*/
    
    util::Future<renderer::Bitmap> bitmap = this->support->getTile(latIndex, lngIndex, tsLevel/*, opts*/).getFuture();
    if(!bitmap.valid()) {
        return TileReader::Error;
    }
    
    try {
        renderer::Bitmap r = bitmap.get();
        
        if(bitmap.getState() == util::SharedState::Canceled)
            return TileReader::Canceled;
        else if(r.data == NULL)
            return TileReader::Error;
        else {
#if DEBUG_TILE_DRAW
            std::stringstream ss;
            ss << level << ":" << tileColumn << ", " << tileRow;
            std::auto_ptr<renderer::BitmapCompositor> compositor(renderer::BitmapCompositor::create(r));
            compositor->debugText(0, 0, r.width, r.height, ss.str().c_str());
#endif
            *retval = r;
            return TileReader::Success;
        }
    } catch (const std::exception &e) {
    }
    return TileReader::Error;
}

LegacyTilePyramidTileReader::ReadResult
LegacyTilePyramidTileReader::read(int level, int64_t tileColumn, int64_t tileRow, void *data, size_t dataSize) {
    
    if (level < 0)
        throw std::invalid_argument("level");
    
    renderer::Bitmap bitmap;
    bitmap.data = data;
    bitmap.dataLen = dataSize;
    bitmap.releaseData = NULL;
    ReadResult code = this->getTileImpl(level, tileColumn, tileRow, &bitmap);
    
    //XXX-- get rid of copy
    if (code == Success) {
        memcpy(data, bitmap.data, bitmap.dataLen);
    }
    
    if (bitmap.releaseData) {
        bitmap.releaseData(bitmap);
    }
    
    return code;
}

struct RectF {
    float top, left, bottom, right;
};

LegacyTilePyramidTileReader::ReadResult
LegacyTilePyramidTileReader::read(int64_t srcX, int64_t srcY, int64_t srcW, int64_t srcH, int dstW,
                int dstH, void *data, size_t dataSize) {
    
    const int level = math::clamp((int)(log(fmax((double)srcW / (double)dstW, (double)srcH / (double)dstH)) / log(2.0)), 0, this->levelCount-1);
    
    const double subsampleX = (double)dstW / ((double)srcW/(double)(1<<level));
    const double subsampleY = (double)dstH / ((double)srcH/(double)(1<<level));
    
    const int64_t stx = math::clamp(this->getTileColumn(level, srcX), 0ll, this->getNumTilesX(level) - 1);
    const int64_t ftx = math::clamp(this->getTileColumn(level, srcX+srcW-1), 0ll, this->getNumTilesX(level)-1);
    
    const int64_t sty = math::clamp(this->getTileRow(level, srcY), 0ll, this->getNumTilesY(level)-1);
    const int64_t fty = math::clamp(this->getTileRow(level, srcY+srcH-1), 0ll, this->getNumTilesY(level)-1);

    // if we are reading exactly one tile, skip the compositing
    if(subsampleX == 1.0 && subsampleY == 1.0 &&
       stx == ftx && sty == fty &&
       dstW == this->tileWidth && dstH == this->tileHeight) {
        
       return this->read(level, stx, sty, data, dataSize);
    }
    
#if RESTRICT_COMPOSITE
    //XXX-- stop if the read is too far out
    int64_t area = (ftx - stx) * (fty - sty);
    if (area > 16) {
        return Error;
    }
#endif
    
    ReadResult result = Success;
    renderer::BitmapOptions bitmapOpts;
    bitmapOpts.dataType = GL_UNSIGNED_BYTE;
    bitmapOpts.format = GL_RGBA;
    bitmapOpts.width = dstW;
    bitmapOpts.height = dstH;
    renderer::Bitmap bitmap = renderer::BitmapFactory::create(bitmapOpts);
    ReadResult code;
    
    std::auto_ptr<renderer::BitmapCompositor> compositor(renderer::BitmapCompositor::create(bitmap));
    renderer::Bitmap tile;

    // XXX-- why are these flipped from Android?
    int64_t aty = fty;

    for(int64_t ty = sty; ty <= fty; ty++) {
        
        //RectF dst;
        float dstY = (float)((double)((this->getTileSourceY(level, ty)-srcY)>>level) * subsampleY);
        float dstH = (float)((double)((this->getTileSourceY(level, ty+1)-1-srcY)>>level) * subsampleY) - dstY;
        
        dstY = floorf(dstY);
        dstH = ceilf(dstH);
        
        for(int64_t tx = stx; tx <= ftx; tx++) {
            
            if (!this->valid) {
                return Error;
            }
            
            code = this->getTileImpl(level, tx, aty, &tile);
            if(code == Success) {
                float dstX = (float)((double)((this->getTileSourceX(level, tx)-srcX)>>level) * subsampleX);
                float dstW = (float)((double)((this->getTileSourceX(level, tx+1)-1-srcX)>>level) * subsampleX) - dstX;
                
                dstX = floorf(dstX);
                dstW = ceilf(dstW);
                
                compositor->composite(dstX, dstY, dstW, dstH,
                                      0, 0, tile.width, tile.height,
                                      tile);
                
                if (tile.releaseData) {
                    tile.releaseData(tile);
                    tile.releaseData = nullptr;
                }
            }
        }
        
        --aty;
    }
    
    memcpy(data, bitmap.data, bitmap.dataLen);
    if (bitmap.releaseData) {
        bitmap.releaseData(bitmap);
    }
    
    return result;
}

/**************************************************************************/

std::string LegacyTilePyramidTileReader::generateKey(renderer::AsyncBitmapLoader *loader, TilesetInfo *info) {

    static uint64_t keyIndex = 1;
    uint64_t value = keyIndex++;

    // XXX - hash codes? static counter?

    std::stringstream ss;
    ss << value;
    return ss.str();
}

PGSC::String LegacyTilePyramidTileReader::registerTilesetInfo(renderer::AsyncBitmapLoader *loader, TilesetInfo *tsInfo) {
    
    TAKErr code(TE_Ok);
    LockPtr lock(NULL, NULL);
    code = Lock_create(lock, staticMutex);
    if (code != TE_Ok)
        throw std::runtime_error
        ("LegacyTilePyramidTileReader::registerTilesetInfo: Failed to acquire mutex");
    
    const std::string retval = generateKey(loader, tsInfo);

    //    registerMap.put(retval, new ReferenceCount<Pair<GLBitmapLoader, WeakReference<TilesetInfo>>>(Pair.create(loader, new WeakReference<TilesetInfo>(info))));
    registerMap.insert(std::pair<PGSC::String, RegisterValue>(retval.c_str(), RegisterValue(loader, tsInfo)));

    return ("tsinfo://" + retval).c_str();
}

void LegacyTilePyramidTileReader::unregisterLayer(TilesetInfo *info) {
    
    TAKErr code(TE_Ok);
    LockPtr lock(NULL, NULL);
    code = Lock_create(lock, staticMutex);
    if (code != TE_Ok)
        throw std::runtime_error
        ("LegacyTilePyramidTileReader::unregisterLayer: Failed to acquire mutex");
    
    /*ReferenceCount<Pair<GLBitmapLoader, WeakReference<TilesetInfo>>> ref = NULL;
    
    Iterator<ReferenceCount<Pair<GLBitmapLoader, WeakReference<TilesetInfo>>>> iter;
    
    iter = register.values().iterator();
    TilesetInfo registerInfo;
    while(iter.hasNext()) {
        ref = iter.next();
        
        registerInfo = ref.value.second.get();
        
        // if the reference was reclaimed, remove
        if(registerInfo == NULL)
            iter.remove();
        if(registerInfo == info) {
            ref.dereference();
            if(!ref.isReferenced())
                iter.remove();
            break;
        }
        
    }*/
}

File: takkernel/engine/src/main/java/com/atakmap/map/layer/raster/tilepyramid/LegacyTilePyramidTileReader.java
package com.atakmap.map.layer.raster.tilepyramid;

import java.lang.ref.WeakReference;
import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;
import java.util.UUID;
import java.util.concurrent.CancellationException;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.Future;

import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.util.Pair;

import com.atakmap.android.maps.graphics.GLBitmapLoader;
import com.atakmap.android.maps.tilesets.TilesetInfo;
import com.atakmap.android.maps.tilesets.TilesetSupport;
import com.atakmap.coremap.log.Log;
import com.atakmap.coremap.maps.coords.GeoPoint;
import com.atakmap.map.layer.feature.geometry.Envelope;
import com.atakmap.map.layer.raster.tilereader.TileReader;
import com.atakmap.map.layer.raster.tilereader.TileReaderFactory.Options;
import com.atakmap.map.layer.raster.tilereader.TileReaderSpi;
import com.atakmap.util.ReferenceCount;

public class LegacyTilePyramidTileReader extends AbstractTilePyramidTileReader {

    public final static TileReaderSpi SPI = new TileReaderSpi() {
        @Override
        public String getName() {
            return "tileset";
        }

        @Override
        public TileReader create(String uri, Options options) {
            if(!uri.startsWith("tsinfo://"))
                return null;
            ReferenceCount<Pair<GLBitmapLoader, WeakReference<TilesetInfo>>> entry = register.get(uri.substring(9));
            if(entry == null)
                return null;
            final TilesetInfo tsInfo = entry.value.second.get();
            if(tsInfo == null)
                return null;

            final TilesetSupport support = TilesetSupport.create(tsInfo, entry.value.first);
            if(support == null)
                return null;

            TileReader.AsynchronousIO io = TileReader.getMasterIOThread();
            if(options != null && options.asyncIO != null)
                io = options.asyncIO;
            
            return new LegacyTilePyramidTileReader(tsInfo, support, io);
        }
        
        @Override
        public boolean isSupported(String uri) {
            return uri.startsWith("tsinfo://");
        }
    };

    private final static String TAG = "LegacyTilePyramidTileReader";
    
    private final static Map<String, ReferenceCount<Pair<GLBitmapLoader, WeakReference<TilesetInfo>>>> register = new HashMap<String, ReferenceCount<Pair<GLBitmapLoader, WeakReference<TilesetInfo>>>>();

    // for layer converters that spit out stuff based on 32bit floats...
    private static final double EPSILON = 0.000000001;
    
    private TilesetSupport support;
    
    private final int gridOffsetX;
    private final int gridOffsetY;
    private final int gridWidth;
    private final int gridHeight;

    private Future<Bitmap> pending;
 
    private static int checkAndThrow(int level) { 
        if (level < 0) 
             throw new IllegalArgumentException("TilesetInfo level count is 0");
        return level;
    }

    private LegacyTilePyramidTileReader(TilesetInfo info, TilesetSupport support, AsynchronousIO io) {
        super(info.getInfo().getUri(),
              null,
              io,
              info.getLevelCount(),
              info.getTilePixelWidth()*(getGridWidth(info)<<(checkAndThrow(info.getLevelCount()-1))),
              info.getTilePixelHeight()*(getGridHeight(info)<<(checkAndThrow(info.getLevelCount()-1))),
              info.getTilePixelWidth(),
              info.getTilePixelHeight());

        this.support = support;
        if(this.support == null)
            throw new NullPointerException("TilesetSupport cannot be null");

        this.support.init();
        this.support.start();
                
        final double zeroHeight = info.getZeroHeight();
        final double zeroWidth = info.getZeroWidth();

        final double gridOriginLat = info.getGridOriginLat();
        final double gridOriginLng = info.getGridOriginLng();

        Envelope mbb = info.getInfo().getCoverage(null).getEnvelope();

        double south = mbb.minY;
        double north = mbb.maxY;
        double west = mbb.minX;
        double east = mbb.maxX;

        // Shrink down on the 4 corners as much as we can
        GeoPoint sw = info.getInfo().getLowerLeft();
        GeoPoint nw = info.getInfo().getUpperLeft();
        GeoPoint ne = info.getInfo().getUpperRight();
        GeoPoint se = info.getInfo().getLowerRight();

        final double coverageSouth = Math.min(sw.getLatitude(), se.getLatitude());
        final double coverageNorth = Math.max(nw.getLatitude(), ne.getLatitude());
        final double coverageEast = Math.max(ne.getLongitude(), se.getLongitude());
        final double coverageWest = Math.min(nw.getLongitude(), sw.getLongitude());

        south = _alignMin(gridOriginLat, Math.max(coverageSouth, south), zeroHeight);
        west = _alignMin(gridOriginLng, Math.max(coverageWest, west), zeroWidth);
        north = _alignMax(gridOriginLat, Math.min(coverageNorth, north), zeroHeight);
        east = _alignMax(gridOriginLng, Math.min(coverageEast, east), zeroWidth);

        int _gridHeight = info.getGridHeight();
        if (_gridHeight < 0)
            _gridHeight = (int) (((north - south) + EPSILON) / zeroHeight);
        int _gridWidth = info.getGridWidth();
        if (_gridWidth < 0)
            _gridWidth = (int) (((east - west) + EPSILON) / zeroWidth);
        int _gridX = info.getGridOffsetX();
        if (_gridX < 0)
            _gridX = (int) ((west - gridOriginLng + EPSILON) / zeroWidth);
        int _gridY = info.getGridOffsetY();
        if (_gridY < 0)
            _gridY = (int) ((south - gridOriginLat + EPSILON) / zeroHeight);

        this.gridHeight = _gridHeight;
        this.gridWidth = _gridWidth;
        this.gridOffsetX = _gridX;
        this.gridOffsetY = _gridY;

        // commenting out this otherwise debug log information
    
        //System.out.println("tileset numlevels=" + this.numAvailableLevels);
        //System.out.println("tileset gridwidth[max]=" + this.getNumTilesX(0));
        //System.out.println("tileset gridheight[max]=" + this.getNumTilesY(0));
        //System.out.println("tileset gridwidth[min]=" + this.getNumTilesX(this.numAvailableLevels-1));
        //System.out.println("tileset gridheight[min]=" + this.getNumTilesY(this.numAvailableLevels-1));
        //System.out.println("tileset width=" + getWidth());
        //System.out.println("tileset height=" + getHeight());
        //System.out.println("tileset gridOffsetX=" + gridOffsetX);
        //System.out.println("tileset gridOffsetY=" + gridOffsetY);
        //System.out.println("tileset gridWidth=" + gridWidth);
        //System.out.println("tileset gridHeight=" + gridHeight);
        //System.out.println("tileset tileWidth=" + tileWidth);
        //System.out.println("tileset tileHeight=" + tileHeight);
     
    }
    
    private static int getGridWidth(TilesetInfo info) {
        int _gridWidth = info.getGridWidth();
        if(_gridWidth < 0) {
            final double zeroWidth = info.getZeroWidth();
    
            final double gridOriginLng = info.getGridOriginLng();
    
            Envelope mbb = info.getInfo().getCoverage(null).getEnvelope();
    
            double west = mbb.minX;
            double east = mbb.maxX;
    
            // Shrink down on the 4 corners as much as we can
            GeoPoint sw = new GeoPoint(mbb.minY, mbb.minX);
            GeoPoint nw = new GeoPoint(mbb.maxY, mbb.minX);
            GeoPoint ne = new GeoPoint(mbb.maxY, mbb.maxX);
            GeoPoint se = new GeoPoint(mbb.minY, mbb.maxX);
    
            final double coverageEast = Math.max(ne.getLongitude(), se.getLongitude());
            final double coverageWest = Math.min(nw.getLongitude(), sw.getLongitude());
    
            west = _alignMin(gridOriginLng, Math.max(coverageWest, west), zeroWidth);
            east = _alignMax(gridOriginLng, Math.min(coverageEast, east), zeroWidth);

            _gridWidth = (int) (((east - west) + EPSILON) / zeroWidth);
        }
        return _gridWidth;
    }

    private static int getGridHeight(TilesetInfo info) {
        int _gridHeight = info.getGridHeight();
        if(_gridHeight < 0) {
            final double zeroHeight = info.getZeroHeight();
    
            final double gridOriginLat = info.getGridOriginLat();
    
            Envelope mbb = info.getInfo().getCoverage(null).getEnvelope();
    
            double south = mbb.minY;
            double north = mbb.maxY;
    
            // Shrink down on the 4 corners as much as we can
            GeoPoint sw = new GeoPoint(mbb.minY, mbb.minX);
            GeoPoint nw = new GeoPoint(mbb.maxY, mbb.minX);
            GeoPoint ne = new GeoPoint(mbb.maxY, mbb.maxX);
            GeoPoint se = new GeoPoint(mbb.minY, mbb.maxX);
    
            final double coverageSouth = Math.min(sw.getLatitude(), se.getLatitude());
            final double coverageNorth = Math.max(nw.getLatitude(), ne.getLatitude());
    
            south = _alignMin(gridOriginLat, Math.max(coverageSouth, south), zeroHeight);
            north = _alignMax(gridOriginLat, Math.min(coverageNorth, north), zeroHeight);

            _gridHeight = (int) (((north - south) + EPSILON) / zeroHeight);

        }
        return _gridHeight;
    }

    @Override
    public void start() {
        this.support.start();
    }
    
    @Override
    public void stop() {
        this.support.stop();
    }
    
    /**************************************************************************/
    // Tile Reader
    
    @Override
    public void disposeImpl() {
        super.disposeImpl();
        if(this.support != null) {
            this.support.release();
            this.support = null;
        }
    }

    @Override
    protected synchronized void cancel() {
        if(this.pending != null)
            this.pending.cancel(true);
    }

    @Override
    protected Bitmap getTileImpl(int level, long tileColumn, long tileRow, ReadResult[] code) {
        return getTileImpl(level, tileColumn, tileRow, null, code);
    }

    @Override
    protected Bitmap getTileImpl(int level, long tileColumn, long tileRow, BitmapFactory.Options opts, ReadResult[] code) {
        if (level < 0)
            throw new IllegalArgumentException();

        // need to invert level and tile row for lower-left origin of legacy
        // tileset infrastructure
        final int tsLevel = (this.numAvailableLevels-level-1);
        tileRow = (this.gridHeight<<tsLevel)-tileRow-1;

        final int latIndex = (int)tileRow + (this.gridOffsetY<<tsLevel);
        final int lngIndex = (int)tileColumn + (this.gridOffsetX<<tsLevel);

        if(opts.inPreferredConfig == null)
            opts.inPreferredConfig = Bitmap.Config.ARGB_8888;

        Future<Bitmap> bitmap = this.support.getTile(latIndex, lngIndex, tsLevel, opts);
        if(bitmap == null) {
            if(code != null) code[0] = ReadResult.ERROR;
            return null;
        }
        
        try {
            while(!bitmap.isDone())
                try {
                    this.wait(20);
                } catch(InterruptedException ignored) {}

            final Bitmap retval = bitmap.get();
            
            if(code != null) {
                if(bitmap.isCancelled())    code[0] = ReadResult.CANCELED;
                else if(retval == null)     code[0] = ReadResult.ERROR;
                else                        code[0] = ReadResult.SUCCESS;
            }
                
            return retval;
        } catch(CancellationException e) {
            if(code != null) code[0] = ReadResult.CANCELED;
            return null;
        } catch(ExecutionException e) {
            if(code != null) code[0] = ReadResult.ERROR;
            Log.w(TAG, "Tile [" + level + "," + tileColumn + "," + tileRow + "] failed to load", e);
            return null;
        } catch(InterruptedException e) {
            if(code != null) code[0] = ReadResult.CANCELED;
            return null;
        }
    }
    
    /**************************************************************************/
    
    private static double _alignMin(double o, double v, double a) {
        double n = (v - o) / a;
        return (Math.floor(n) * a) + o;
    }

    private static double _alignMax(double o, double v, double a) {
        double n = (v - o) / a;
        return (Math.ceil(n) * a) + o;
    }
    
    /**************************************************************************/
    
    private static String generateKey(GLBitmapLoader loader, TilesetInfo info) {
        // XXX - hash codes? static counter?
        return UUID.randomUUID().toString();
    }
    public synchronized static String registerTilesetInfo(GLBitmapLoader loader, TilesetInfo info) {
        final String retval = generateKey(loader, info);
        register.put(retval, new ReferenceCount<Pair<GLBitmapLoader, WeakReference<TilesetInfo>>>(Pair.create(loader, new WeakReference<TilesetInfo>(info))));
        return "tsinfo://" + retval;
    }
    
    public synchronized static void unregisterLayer(TilesetInfo info) {
        ReferenceCount<Pair<GLBitmapLoader, WeakReference<TilesetInfo>>> ref = null;
        
        Iterator<ReferenceCount<Pair<GLBitmapLoader, WeakReference<TilesetInfo>>>> iter;
        
        iter = register.values().iterator();
        TilesetInfo registerInfo;
        while(iter.hasNext()) {
            ref = iter.next();
            
            registerInfo = ref.value.second.get();
            
            // if the reference was reclaimed, remove
            if(registerInfo == null)
                iter.remove();
            if(registerInfo == info) {
                ref.dereference();
                if(!ref.isReferenced())
                    iter.remove();
                break;
            }
            
        }
    }
}
